콜백 함수

비동기 함수를 쓰기 위해서 callback함수를 쓴다

```jsx
function loadScript(src, callback){
	let script = document.createElement('script');
	script.src = src;

	script.onload = () => callback(script);

	document.head.append(script);
}
```

새롭게 불러온 스크립트에 있는 함수를 콜백함수 안에서 호출하면 원하는 대로 외부 스크립트 안의 함수를 사용할 수 있습니다.

```jsx
loadScript('/my/script.js', function(){
	//콜백 함수는 스크립트 로드가 끝나면 실행됩니다.
	newFunction();
	...
});
```

이렇게 두 번째 인수로 전달된 함수는 원하는 동작(위 예제에선 외부 스크립트를 불러오는 것)이 완료되었을 때 실행됩니다.

아래는 실제 존재하는 스크립트를 이용해 만든 예시이다

```jsx
function loadScript(src, callback){
	let script = document.createElement('script');
	script.src = src;
	script.onload = () => callback(script);
	document.head.append(script);
}

loadScript('<https://cdnjs.cloudflare.com/ajax/libs/lodash.js/3.2.0/lodash.js>', script => {
  alert(`${script.src}가 로드되었습니다.`);
  alert( _ ); // 스크립트에 정의된 함수
});
```

이런 방식을 ‘콜백 기반’ 비동기 프로그래밍이라고 합니다. 무언가를 비동기적으로 수해하는 함수는 함수 내 동작이 모두 처리된 후 실행되어야 하는 함수가 들어갈 `콜백` 을 인수로 반드시 제공해야 합니다.

위 예시에선 `loadScript` 의 인수로 콜백을 제공해 주었는데, 이렇게 콜백을 사용한 방식은 비동기 프로그래밍의 일반적인 접근법입니다.

콜백 속 콜백 같은 것이 가능하다.

```jsx
loadScript('/my/script.js', function(script) {

  alert(`${script.src}을 로딩했습니다. 이젠, 다음 스크립트를 로딩합시다.`);

  loadScript('/my/script2.js', function(script) {
    alert(`두 번째 스크립트를 성공적으로 로딩했습니다.`);
  });

});
```

promise

```jsx
let promise = new Promise(function(resolve, reject){
	//executor( 제작코드, '가수')
});
```

* resolve(value) - 일이 성공적으로 끝난 경우 그 결과를 나타내는 value와 함께 호출
* reject(error) - 에러 발생 시 에러 객체를 나타내는 error와 함께 호출

executor는 자동으로 실행되는데 여기서 원하는 일이 처리됩니다. 처리가 끝나면 executor는 처리 성공 여부에 따라 resolve나 reject를 호출합니다.

한편 new Promise 생성자가 반환하는 promise 객체는 다음과 같은 내부 프로퍼티를 갖습니다.

* state - 처음엔 “pending”(보류)이었다 resolve가 호출되면 “fullfilled”, reject가 호출되면 “rejected”로 변합니다
* result - 처음엔 “undefined”이었다 resolve(Value)가 호출되면 value로, reject(error)가 호출되면 error로 변합니다

```jsx
let promise = new Promise(function(resolve, reject){
	//프라미스가 만들어지면 executor 함수는 자동으로 실행되고
	//1초 뒤에 일이 성공적으로 끝났다는 신호가 전달되면서 result는 '완료'가 됩니다.
	setTimeout(() => resolve("완료"), 1000);
});
```

```jsx
let promise = new Promise(function(resolve, reject){
	//프라미스가 만들어지면 executor 함수는 자동으로 실행되고
	//1초 뒤에 일이 성공적으로 끝났다는 신호가 전달되면서 result는 '완료'가 됩니다.
	setTimeout(() => reject(new Error("에러 발생!")), 1000);
});
```

then

.then은 프라미스에서 가장 중요하고 기본이 되는 메서드입니다.

문법

```jsx
promise.then(
	function(result){결과를 다룬다}
	function(error){에러를 다룬다}
};
```

예시

```jsx
let promise = new Promise(function(resolve, reject){
	setTimeout(()=> resolve("완료"),1000);
});

promise.then(
	result => alert(result),
	error => alert(error)
};
```

프라미스가 거부된 경우에는 아래와 같이 두 번째 함수가 실행됩니다.

```jsx
let promise = new Promise(function(resolve, reject) {
  setTimeout(() => reject(new Error("에러 발생!")), 1000);
});

// reject 함수는 .then의 두 번째 함수를 실행합니다.
promise.then(
  result => alert(result), // 실행되지 않음
  error => alert(error) // 1초 후 "Error: 에러 발생!"을 출력
);
```

catch와 finally가 있다

```jsx
new Promise((resolve, reject) => {
  setTimeout(() => resolve("결과"), 2000)
})
  .finally(() => alert("프라미스가 준비되었습니다."))
  .then(result => alert(result)); // <-- .then에서 result를 다룰 수 있음

new Promise((resolve, reject) => {
  throw new Error("에러 발생!");
})
  .finally(() => alert("프라미스가 준비되었습니다."))
  .catch(err => alert(err)); // <-- .catch에서 에러 객체를 다룰 수 있음
```
